pydantic/_internal/_core_utils.py

```diff
@@ -416,92 +435,122 @@ def walk_core_schema(schema: core_schema.CoreSchema, f: Walk) -> core_schema.Cor
 
     def count_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
+        if 'metadata' in s:
+            definitions_cache: _DefinitionsState | None = s['metadata'].get(_DEFINITIONS_CACHE_METADATA_KEY, None)
+            if definitions_cache is not None:
+                for ref in definitions_cache['ref_counts']:
+                    state['ref_counts'][ref] += definitions_cache['ref_counts'][ref]
+                    # it's possible that a schema was seen before we hit the cache
+                    # and also exists in the cache, in which case it is involved in a recursion
+                    if state['current_recursion_ref_count'][ref] != 0:
+                        state['involved_in_recursion'][ref] = True
+                # if it's involved in recursion in the inner schema mark it globally as involved in a recursion
+                for ref_in_recursion in definitions_cache['involved_in_recursion']:
+                    if ref_in_recursion:
+                        state['involved_in_recursion'][ref_in_recursion] = True
+                return s
+
         if s['type'] != 'definition-ref':
             return recurse(s, count_refs)
         ref = s['schema_ref']
-        ref_counts[ref] += 1
+        state['ref_counts'][ref] += 1
 
-        if ref_counts[ref] >= 2:
+        if state['ref_counts'][ref] >= 2:
             # If this model is involved in a recursion this should be detected
             # on its second encounter, we can safely stop the walk here.
-            if current_recursion_ref_count[ref] != 0:
-                involved_in_recursion[ref] = True
+            if state['current_recursion_ref_count'][ref] != 0:
+                state['involved_in_recursion'][ref] = True
             return s
 
-        current_recursion_ref_count[ref] += 1
-        recurse(all_defs[ref], count_refs)
-        current_recursion_ref_count[ref] -= 1
+        state['current_recursion_ref_count'][ref] += 1
+        recurse(state['definitions'][ref], count_refs)
+        state['current_recursion_ref_count'][ref] -= 1
         return s
 
     schema = walk_core_schema(schema, count_refs)
 
-    assert all(c == 0 for c in current_recursion_ref_count.values()), 'this is a bug! please report it'
+    assert all(c == 0 for c in state['current_recursion_ref_count'].values()), 'this is a bug! please report it'
+
+    definitions_cache = _DefinitionsState(
+        definitions=state['definitions'],
+        ref_counts=dict(state['ref_counts']),
+        involved_in_recursion=state['involved_in_recursion'],
+        current_recursion_ref_count=dict(state['current_recursion_ref_count']),
+    )
+
+    def can_be_inlined(s: core_schema.DefinitionReferenceSchema, ref: str) -> bool:
+        if state['ref_counts'][ref] > 1:
+            return False
+        if state['involved_in_recursion'].get(ref, False):
+            return False
+        if 'serialization' in s:
+            return False
```

Comment:
What's the motivation for this check? Seems somewhat surprising to me that serialization override prevents inlining.

Reply:
We use this to modify serialization on a type without globally modifying a type. That is, you can wrap a type with a serializer